WordPress/gutenberg
# Preserve bindings metadata in block transforms

## What?
Fixes https://github.com/WordPress/gutenberg/issues/59086

This pull request adds the possibility of preserving and adapting the metadata attribute for the following transforms:
* Paragraph to heading.
* Heading to paragraph.
* Paragraph to button.

I've also added some tests to check that, in those transformations, the content, align, and metadata attributes are preserved. I couldn't find any existing tests for that. Please let me know if that's not the case.

A couple of questions that come to my mind:
* Should we preserve any data stored in `metadata` or only the bindings?
* Should we explore the possibility of using a hook to make it work for all transforms instead of hardcoding it for these ones?

## Why?
Right now, if you rename your block, or add a binding to it, and transform it into another block, that data is not preserved.

## How?
I just added some logic in the heading, and the button transforms objects to pass the `metadata` attribute as well.

## Testing Instructions

**Test renaming a block is preserved**

* Go to a page and add a paragraph.
* Click on the block options and rename it.
* Transform paragraph into a heading.
* Check that the custom name.

Repeat the process transforming a heading into a paragraph and a paragraph into a button.

**Test bindings in a block are preserved**

* Go to a page and add a paragraph.
* Go to the Code editor and add a binding to it.
```
{"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"my_custom_field"}}}}}
```
* Go to the Visual editor and transform the paragraph into a heading.
* Check that the bindings object is preserved.

Repeat the process transforming a heading into a paragraph.

Repeat the process transforming a paragraph into a button. In this case, the object should change to point to the "text" attribute instead of "content".
